src/system_check.cc

```diff
@@ -0,0 +1,86 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+#include "system_checks.hpp"
+
+#include "chrono_utils.hpp"
+#include "defer.hpp"
+#include "logger.hpp"
+#include "tsc_clock.hpp"
+#include "unique_fd.hpp"
+
+#include <chrono>
+#include <fstream>
+#include <optional>
+#include <sys/resource.h>
+#include <unistd.h>
+
+namespace ddprof {
+
+namespace {
+
+constexpr auto kCurrentClockSourceSysFsPath =
+    "/sys/devices/system/clocksource/clocksource0/current_clocksource";
+constexpr auto kAvailableClockSourcesSysFsPath =
+    "/sys/devices/system/clocksource/clocksource0/available_clocksource";
+
+std::optional<std::string> read_line_from_file(const char *path) {
+  std::ifstream ifs(path);
+  std::string line;
+  std::getline(ifs, line);
+  return ifs ? std::optional{line} : std::nullopt;
+}
+
+void check_clock_source() {
+  auto line = read_line_from_file(kCurrentClockSourceSysFsPath);
+  if (line == "xen") {
+    LG_WRN("xen clock source detected. This might lead to degraded "
+           "performance.");
+  }
+}
+
```

Comment:
Interesting function!
A small comment around the strategy would be interesting.
Idea for the backlog: This is something we could report as a tag within the reported profiles.

Reply:
Added some comments